src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,161 @@
+#include <sc/sidechainTxsCommitmentBuilder.h>
+#include <primitives/transaction.h>
+#include <primitives/certificate.h>
+
+#ifdef BITCOIN_TX
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx) { return; }
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert) { return; }
+uint256 SidechainTxsCommitmentBuilder::getCommitment() { return uint256(); }
+#else
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx)
+{
+    if (!tx.IsScVersion())
+        return;
+
+    unsigned int ccOutIdx = 0;
+    LogPrint("sc", "%s():%d -getting leaves for vsc out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVscCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVscCcOut().at(ccOutIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVscCcOut()[ccOutIdx].GetHash(), tx.GetHash(), ccOutIdx);
+        vec.push_back(pField);
+    }
+
+    LogPrint("sc", "%s():%d -getting leaves for vft out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVftCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVftCcOut().at(ccOutIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVftCcOut()[ccOutIdx].GetHash(), tx.GetHash(), ccOutIdx);
+        vec.push_back(pField);
+    }
+
+    LogPrint("sc", "%s():%d - nIdx[%d]\n", __func__, __LINE__, ccOutIdx);
+}
+
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert)
+{
+    uint256 scId = cert.GetScId();
+    sScIds.insert(scId);
+    mScCerts[scId] = mapCertToField(cert.GetHash());
+}
+
+uint256 SidechainTxsCommitmentBuilder::getCommitment()
+{
+    std::vector<field_t*> vSortedScLeaves;
+
+    // set of scid is ordered
+    for (const auto& scid : sScIds)
+    {
+        field_t* ftRootField = nullptr;
+        auto itFt = mScMerkleTreeLeavesFt.find(scid);
+        if (itFt != mScMerkleTreeLeavesFt.end() )
+        {
+            ftRootField = merkleTreeRootOf(itFt->second);
+        }
+
+        field_t* btrRootField = nullptr;
+        auto itBtr = mScMerkleTreeLeavesBtr.find(scid);
+        if (itBtr != mScMerkleTreeLeavesBtr.end() )
+        {
+            btrRootField = merkleTreeRootOf(itBtr->second);
+        }
+
+        field_t* certRootField = nullptr;
+        auto itCert = mScCerts.find(scid);
+        if (itCert != mScCerts.end() )
+        {
+            certRootField = itCert->second;
+        }
+
+        field_t* scIdField = mapCertToField(scid);
+
+        std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
+        field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
```

Comment:
Current solution is different to the previous one. Previously we calculated at first the merkle tree root for `{ftRootField, btrRootField}` -> `txsHash`. Then, we did it once more for `{txsHash, certRootField, scIdField}`
Honestly, I don't remember exactly the reason, but seems that such order was important (see page 17 in zendoo paper).

Reply:
Indeed I changed wrt to previous implementation. I made it slightly closer to what @DanieleDiBenedetto told me the final implementation would be (with fix number of leaves for fwds, btrs and certs and number of sidechains too).
Bear in mind that this is not yet an "active" pull request, and I made this only to have early feedbacks and ease up the following changes to our crypto libs.
Moreover, since current validations are only in a separate testnet, we may even have no need to ensure backward compatibility.